publish-mqtt/src/main.rs

```diff
@@ -187,77 +191,150 @@ impl Sensor {
+    homie
+        .ready()
+        .await
+        .with_context(|| std::line!().to_string())?;
+
+    let state = Arc::new(Mutex::new(SensorState {
+        sensors_to_connect: VecDeque::new(),
+        sensors_connected: vec![],
+        homie,
+    }));
+
+    let mut next_scan_due = Instant::now();
+
+    let t1 = async {
+        loop {
+            let now = Instant::now();
+            if now > next_scan_due {
+                next_scan_due = now + SCAN_INTERVAL;
+                check_for_sensors(state.clone(), dbus_conn.clone(), &sensor_names)
+                    .await
+                    .with_context(|| std::line!().to_string())?;
+            }
 
-    loop {
-        connect_first_sensor_in_queue(
-            &bt_session,
-            &mut homie,
-            &mut sensors_connected,
-            &mut sensors_to_connect,
-        )
-        .await?;
+            {
+                let state = &mut *state.lock().await;
+                connect_first_sensor_in_queue(
+                    dbus_conn.clone(),
+                    &mut state.homie,
+                    &mut state.sensors_connected,
+                    &mut state.sensors_to_connect,
+                )
+                .await
+                .with_context(|| std::line!().to_string())?;
+            }
 
-        disconnect_first_stale_sensor(&mut homie, &mut sensors_connected, &mut sensors_to_connect)
-            .await?;
+            {
+                let state = &mut *state.lock().await;
+                disconnect_first_stale_sensor(
+                    &mut state.homie,
+                    &mut state.sensors_connected,
+                    &mut state.sensors_to_connect,
+                )
+                .await
+                .with_context(|| std::line!().to_string())?;
+            }
+            time::delay_for(CONNECT_INTERVAL).await;
+        }
+        #[allow(unreachable_code)]
+        Ok(())
+    };
+    let t2 = async {
+        service_bluetooth_event_queue(state.clone(), dbus_conn.clone())
+            .await
+            .with_context(|| std::line!().to_string())?;
+        Ok(())
+    };
+    try_join!(t1, t2).map(|((), ())| ())
+}
 
-        service_bluetooth_event_queue(
-            &bt_session,
-            &mut homie,
-            &mut sensors_connected,
-            &mut sensors_to_connect,
-        )
-        .await?;
+async fn check_for_sensors(
+    state: Arc<Mutex<SensorState>>,
+    dbus_conn: Arc<SyncConnection>,
+    sensor_names: &HashMap<String, String>,
+) -> Result<(), anyhow::Error> {
+    let adapter = dbus::nonblock::Proxy::new(
+        "org.bluez",
+        "/org/bluez/hci0",
+        Duration::from_secs(30),
+        dbus_conn.clone(),
+    );
+    adapter
+        .set_powered(true)
+        .await
+        .with_context(|| std::line!().to_string())?;
+    adapter
+        .start_discovery()
+        .await
+        .unwrap_or_else(|err| println!("starting discovery failed {:?}", err));
```

Comment:
- [ ] make this error println sound less scary (we expect it to explode every time other than the first time we call it)